PowerPointLabs/PowerPointLabs/NarrationsLab/ViewModel/Synthesize.cs

```diff
@@ -0,0 +1,255 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Windows;
+using System.Xml.Linq;
+using PowerPointLabs.ActionFramework.Common.Log;
+using PowerPointLabs.NarrationsLab.Data;
+
+namespace PowerPointLabs.NarrationsLab.ViewModel
+{
+    public class Synthesize
+    {
+
+        private HttpClient client;
+        private HttpClientHandler handler;
+
+        public Synthesize()
+        {
+            var cookieContainer = new CookieContainer();
+            handler = new HttpClientHandler() { CookieContainer = new CookieContainer(), UseProxy = false };
+            client = new HttpClient(handler);
+        }
+
+        ~Synthesize()
+        {
+            client.Dispose();
+            handler.Dispose();
+        }
+
+        public event EventHandler<GenericEventArgs<Stream>> OnAudioAvailable;
+
+        public event EventHandler<GenericEventArgs<Exception>> OnError;
+
+        public Task Speak(CancellationToken cancellationToken, InputOptions inputOptions, string filepath)
+        {
+            client.DefaultRequestHeaders.Clear();
+            foreach (var header in inputOptions.Headers)
+            {
+                client.DefaultRequestHeaders.TryAddWithoutValidation(header.Key, header.Value);
+            }
+
+            var genderValue = "";
+            switch (inputOptions.VoiceType)
+            {
+                case Gender.Male:
+                    genderValue = "Male";
+                    break;
+
+                case Gender.Female:
+                default:
+                    genderValue = "Female";
+                    break;
+            }
+
+            var request = new HttpRequestMessage(HttpMethod.Post, inputOptions.RequestUri)
+            {
+                Content = new StringContent(GenerateSsml(inputOptions.Locale, genderValue, inputOptions.VoiceName, inputOptions.Text))
+            };
+
+            var httpTask = client.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken);
+
+            for (int i = 0; i < 3 && !httpTask.Result.IsSuccessStatusCode; i++)
+            {
+                Thread.Sleep(20000);
+                request = new HttpRequestMessage(HttpMethod.Post, inputOptions.RequestUri)
+                {
+                    Content = new StringContent(GenerateSsml(inputOptions.Locale, genderValue, inputOptions.VoiceName, inputOptions.Text))
```

Comment:
Could this code be placed in a function considering that it is a duplicate of lines 59 to 64?

Reply:
Thanks for the suggestion! Refactored to a separate method.